### PR TITLE
Consistently use (deprecated but `scikit-learn`-compatible) `numpy.random.RandomState` for random number generation

### PR DESCRIPTION
**Pull request by mcw92**
This PR addresses two major things, i.e., (1) (hopefully) improved random number generation throughout `specialcouscous` and (2) introduction of a new accuracy metric and improved distinction between the different accuracy metrics calculated.

## Random number generation

Initially, we used plain integer seeds, one for data generation and splitting (must be the same for each rank to ensure consistent generation of the same global dataset on each rank) and one for setting up each rank-local sub forest (must be different for each rank to ensure different rank-local classifiers). These integers would be passed down to all functions and classes in `specialcouscous` using them.

### Model-related random number generation
The model base seed was passed to the `DistributedRandomForest` class. In the constructor, a rank-local integer seed was constructed by adding the rank to the user-provided base seed. This rank-local integer was then used to set up the local classifier on each rank.

### Data-related random number generation
The seeds for data generation and splitting were passed to the related functions for dataset generation and splitting in the `SyntheticDataset` class. In the balanced case, the data generation integer seed was passed to `sklearn`'s `make_classification` function and the data splitting seed to `sklearn`'s `train_test_split` function as the `random_state`. Internally, each `sklearn` function taking a `random_state` converts a provided integer seed into a [deprecated `numpy.random.RandomState` object](https://numpy.org/doc/stable/reference/random/legacy.html), which is then used for random number generation.
In the imbalanced case, the original data generation seed would be used for both dataset generation and splitting. Internally, this seed would either be passed further to an `sklearn` function, using it to seed a `RandomState` object, or to in-house methods of `SyntheticDataset`, using it to seed a `Generator` (this is how it should be done in modern `numpy`). However, `sklearn` cannot handle generators so far (also check out https://github.com/scikit-learn/scikit-learn/issues/16988, an open issue for more than four years 🙈), which is why I chose to re-organize the random number generation using `RandomState` objects throughout even though one normally should definitely prefer generators.

Before setting up the dropping accuracy experiments, I wanted to clean up and better structure the random number generation throughout the whole `specialcouscous` package by changing the following things:
- Exactly two seeds, i.e., a data-related seed and a model-related (base) seed, can be specified. The separate seed for dataset splitting has been removed. In the serial case, only one seed is needed for everything as we have only one model (and not many different rank-local models). Thus, the data-related seed is used throughout and the model-related seed is ignored for serial RF training.
- The data-related seed (the same on each rank!) is directly transformed into a (deprecated) `numpy.random.RandomState` object which is used consistently for data generation, splitting, and partitioning in the ``SyntheticDataset`` class instead of a mixture of integer seeds for ``sklearn`` and generators for in-house methods. 
- The model-related integer seed is passed down to the constructor of the `DistributedRandomForest` class where a rank-local integer seed is constructed by adding the rank to the user-provided base seed. This rank-local integer seeds a rank-local `RandomState` instance which is used to set up the local classifier on each rank.

I hope that these modifications clean up the whole random number generation issue and help understanding what is going on with the dropping accuracy in our experiments. I had a look at the [`RFClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) source code in `sklearn` and it seems like it will not be as easy as just identifying a "seed distance" to avoid redundancy in the subforests. Internally, an integer seed is always directly converted into a `RandomState` via [`sklearn.utils.validation. check_random_state`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html), which returns a `RandomState` instance based on a provided seed parameter of type `RandomState`, `int`, or `None`. In the best case, the suggested changes already resolve the dropping accuracy problem (which I definitely do not expect but let's hope for the best because I am lacking ideas for explanation otherwise 😬).

Also see: 
https://scikit-learn.org/stable/glossary.html#term-random_state
https://scikit-learn.org/stable/common_pitfalls.html#randomness

## Accuracy metrics

After the change to `RandomState` for random number generation, some of the tested parameter combinations in the breaking IID test suddenly failed. I figured that there was still a bug in the evaluation part of the code. I do not know why that only became evident now (because it should not have worked before), but maybe we were just lucky with our seeds.
The problem was for detailed evaluation on the train set without a global model. In this case, each rank would first calculate an array of the local distributions of the predicted classes for each sample in the local evaluation dataset (in this case, the local train dataset). Afterwards, these arrays would be all-gathered. However, this only makes sense if all ranks share the same evaluation dataset (which is why we require either the global model or the test set to be shared). Unfortunately, I did not think about this again when running the evaluation on the train set. When partitioning the training data, each rank holds an exclusive subset of the global train set, and all-gathering prediction results does not make any sense at all. I only noticed this because my communication did not work anymore as not only the train samples differ from rank to rank but also their number can be different for imbalanced datasets, leading to a mismatch with the buffer sizes. Thinking further about this, I noticed that there are not only two but three possible accuracy metrics one could look at for a global model with private test sets:
- Accuracy of each local model on the local dataset: `self.acc_local`
- Accuracy of the global model on the local dataset: `self.acc_global_local`
- Accuracy of the global model on the global dataset: `self.acc_global`

If the global model is not shared but the test set is the same on each rank, the latter two are expected to be the same. I fixed the bug in the code and introduced this additional accuracy metric.

Sorry for writing so much, I hope this is not too confusing. We can also just talk about this and I can answer any questions when you are back in the office. 

I would appreciate your feedback on these two things very much as a sanity check before running the first experiments to investigate the dropping accuracy ☺️.

### CODE REVIEW

**Comment by mcw92** [comment]
### Update
- I further simplified the seed handling. It is now sufficient to pass a single seed only. If no model base seed is provided by the user, this seed will be used to create a `numpy.random.RandomState` instance (the same on each rank), which is used to create a random integer as the model base seed (and further for dataset generation, splitting, and partitioning). 
- I added an alterative option to generate the rank-local model seeds. In the `DistributedRandomForest` constructor, the provided model base seed is used to seed a `RandomState` instance (the same on each rank). This `RandomState` instance is further used to create a sequence of `self.comm.size` random integers. Each rank then uses the value of this sequence at position `self.comm.rank` to seed a rank-local `RandomState` instance which is passed to the rank-local random forest classifier.
- I committed the Python scripts to generate the job scripts for the dropping accuracy experiment series.

File: specialcouscous/rf_parallel.py [hunk 1]
```diff
@@ -311,47 +363,74 @@ def train(
...
+            # Note that this metric can only be calculated if the global model is shared among all ranks.
+            self.acc_global_local = (targets == majority_vote).mean()
+
+        else:  # Global model is not shared. Note that the dataset to be tested must be shared among all ranks.
             # Get class predictions of sub estimators in each forest.
             log.info(f"[{rank}/{size}]: Get predictions of individual sub estimators.")
-            tree_predictions_local = self._predict_locally(test_samples)
+            tree_predictions_local = self._predict_locally(samples)
             log.info(f"[{rank}/{size}]: Calculate majority vote via histograms.")
             predicted_class_hists = self._predicted_class_hist(
```
Comment (fluegelk) [inline specialcouscous/rf_parallel.py:419]: Should we make sure the the samples have the same length here so we can raise an Exception and exit more gracefully instead of failing during the communication?

File: specialcouscous/rf_parallel.py [hunk 2]
```diff
@@ -311,47 +363,74 @@ def train(
...
                 tree_predictions_local, n_classes
             )
             majority_vote = self._calc_majority_vote_hist(predicted_class_hists)
-        # Calculate accuracies.
-        self.acc_global = (test_targets == majority_vote).mean()
-        self.acc_local = self.clf.score(test_samples, test_targets)
+            # Calculate global accuracy as accuracy of global model on local dataset which must be the same on each rank
+            # if the model is not shared globally. The accuracy of the global model on the global dataset is not
+            # accessible for this configuration and set to NaN. It should equal the accuracy of the global model on the
+            # local dataset as the local dataset is the same as the global dataset.
```
Comment (fluegelk) [inline specialcouscous/rf_parallel.py:426]: I find the usage of local vs global dataset still a bit confusing in this comment. Right now, I know what we are trying to say but I fear it might not be clear if I were to return to this code in a year. 

An (unpolished) alternative suggestion:
> Calculate the accuracy of the global model. Since the global model is distributed across the ranks, it can only be evaluated by processing the same data on each ranks local model and aggregating the majority vote across the ranks.
> Evaluating the (distributed) global model is thus only possible if all ranks use the same "local" data for the evaluation (i.e. local data == global data). Evaluating the global model on true local data (not shared with other ranks) is not possible. We thus set acc_global_local to NaN.

**Review by fluegelk — COMMENTED** [review]
I did leave a few minor comments and questions throughout the code, mostly for clarification. Other than that it looks good to me.

File: specialcouscous/train.py [hunk 3]
```diff
@@ -367,8 +370,8 @@ def train_parallel_on_synthetic_data(
     globally_balanced: bool,
     locally_balanced: bool,
     shared_test_set: bool,
-    seed_data: int = 0,
-    seed_model: int = 0,
+    random_state_data: int | np.random.RandomState = 0,
+    random_state_model: int = 0,
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py]: I decided to allow only ints here as it makes the internal handling of the different `RandomState` objects eventually required for different ranks in the `DistributedRandomForest` class easier. For the data stuff, each rank needs the same seed / `RandomState` object which makes it easy to further process a `RandomState` object passed here. For the model, the original workflow took the model base seed, added the local rank, and turned it into a rank-specific `RandomState` object which would not work the same if a `RandomState` object would be passed here. If we would allow passing a `RandomState` object for the model as well, I could use this to generate either the model base seed (in the case of local model seed = model base seed + rank) or the sequence of `comm.size` integers used as local seeds. Shall I implement that, what do you think? Should not be too difficult but not sure whether this is necessary.

File: specialcouscous/train.py [hunk 4]
```diff
@@ -542,9 +564,8 @@ def train_parallel_on_balanced_synthetic_data(
     n_clusters_per_class: int,
     frac_informative: float,
     frac_redundant: float,
-    seed_data: int = 0,
-    seed_split: int = 0,
-    seed_model: int = 0,
+    random_state_data: int | np.random.RandomState = 0,
+    random_state_model: int = 0,
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py]: See answer above ☺️.

File: tests/test_rf_training_breaking_iid.py [hunk 5]
```diff
@@ -15,9 +15,18 @@
 @pytest.mark.parametrize(
     "global_model, private_test_set, globally_imbalanced, locally_imbalanced",
```
[the remark this answers is not in the document]
Comment (mcw92) [reply tests/test_rf_training_breaking_iid.py:16]: Yes, we should do that and maybe I can even do that in this PR.

File: specialcouscous/train.py [hunk 6]
```diff
@@ -256,12 +262,8 @@ def train_serial_on_synthetic_data(
         The fraction of informative features in the dataset.
     frac_redundant : float
         The fraction of redundant features in the dataset.
-    seed_data : int
-        The random seed used for the dataset generation.
-    seed_split : int
-        The random seed used to train-test split the data.
-    seed_model : int
-        The random seed used for the model.
+    random_state : int | np.random.RandomState
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py:265]: I can add that back to enable testing different model seeds for the same dataset.

File: specialcouscous/train.py [hunk 7]
```diff
@@ -558,6 +579,10 @@ def train_parallel_on_balanced_synthetic_data(
     """
     Train and evaluate a distributed random forest on globally balanced synthetic data.
 
+    Note that training and test data are not distributed over the ranks but each rank sees the full dataset. Thus, the
+    test set on each rank is the same even if ``private_test_set`` is set to True as a ``RandomState`` object with the
+    same seed is used for data generation and splitting on each rank.
+
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py:593]: The first half is still true but the second is not needed here and does not help with understanding what is going on. I will change that.

File: specialcouscous/rf_parallel.py [hunk 8]
```diff
@@ -311,47 +363,74 @@ def train(
...
         n_classes : int
             The number of classes in the dataset.
         global_model : bool
-            Whether the global model shall be shared among all ranks (True) or not (False).
+            Whether the global model is shared among all ranks (True) or not (False). Default is True.
         """
         rank, size = self.comm.rank, self.comm.size
-        if global_model:
-            tree_predictions = self._predict_tree_by_tree(test_samples)
+        if global_model:  # Global model is shared.
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/rf_parallel.py:388]: What do you think about `shared_global_model`? Similar to `shared_test_set`?

File: specialcouscous/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fluegelk) [reply specialcouscous/train.py]: That makes sense, I would say it's fine to leave it as is then

File: specialcouscous/rf_parallel.py (hunk 1 above)
Comment (mcw92) [reply specialcouscous/rf_parallel.py:419]: You mean allgathering the local lenghts of the samples and checking whether they are the same before continuing? We can do that as a sanity check but that would not fully prevent "misuse" of this function as it would still work technically if the local datasets to run evalation on have the same number of samples but the samples itself are different. That could, e.g., occur when only distributing the dataset over the ranks without imbalance in the case `num_samples % num_ranks` is 0.

File: specialcouscous/rf_parallel.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fluegelk) [reply specialcouscous/rf_parallel.py:388]: Sounds good, makes sense from a distributed computing context, just like distributed vs shared memory etc

File: specialcouscous/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py]: I will create an issue with very low urgency. Might be something to implement later on.

File: specialcouscous/rf_parallel.py (hunk 1 above)
Comment (fluegelk) [reply specialcouscous/rf_parallel.py:419]: Exactly, it would be primarily to catch the truncated message errors and give a more useful error message in that case.

File: specialcouscous/train.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py:593]: Fixed.

File: specialcouscous/rf_parallel.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/rf_parallel.py]: I changed the variable name throughout the whole code base.

File: tests/test_rf_training_breaking_iid.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcw92) [reply tests/test_rf_training_breaking_iid.py]: I changed `private_test_set` to `shared_test_set` throughout the whole codebase.

File: specialcouscous/rf_parallel.py (hunk 2 above)
Comment (mcw92) [reply specialcouscous/rf_parallel.py:426]: I replaced the comment by your suggestion.

File: specialcouscous/train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/train.py:265]: Done.

File: specialcouscous/rf_parallel.py (hunk 1 above)
Comment (mcw92) [reply specialcouscous/rf_parallel.py:425]: Done.
